sklearn/metrics/cluster/unsupervised.py

```diff
@@ -212,9 +212,10 @@ def silhouette_samples(X, labels, metric='euclidean', **kwds):
     """
     X, labels = check_X_y(X, labels, accept_sparse=['csc', 'csr'])
 
-    # Check for diagonal entries in precomputed distance matrix
+    # Check for non-zero diagonal entries in precomputed distance matrix
     if metric == 'precomputed':
-        if np.any(np.diagonal(X)):
+        atol = np.finfo(X.dtype).eps * 100
```

Comment:
100 seems strange, how do we do similar things in other places?

Reply:
I don't think so, it's pretty arbitrary. What would you suggest?

Reply:
Sorry, misread your comment.

I don't think we have a standard way of dealing with numerical thresholds. I know we sometimes divide by `finfo.eps` to avoid `ZeroDivisionError` e.g. in the gradient boosting code but that's not really the same problem here.

Here I think the main goal is to avoid unexpected test failures like these @rth  mentioned in the other PR.

Reply:
we can't use np.isclose, right?

Reply:
Sure, this should be equivalent to `np.allclose(diag, 0, atol=atol)`